### PR TITLE
Adding SpotPlacementRecommender API cmdlets - version 2024-03-01-preview

### PR DESCRIPTION
**Pull request by Elina-L**
## Description

Adding two new cmdlets for the following APIs:

1. Get-AzSpotPlacementRecommender
2. Invoke-AzSpotPlacementRecommender

## Mandatory Checklist

- Please choose the target release of Azure PowerShell
  - [x] [General release](../blob/main/CONTRIBUTING.md#target-release-types)
  - [ ] [Public preview](../blob/main/CONTRIBUTING.md#target-release-types)
  - [ ] [Private preview](../blob/main/CONTRIBUTING.md#target-release-types)
  - [ ] [Engineering build](../blob/main/CONTRIBUTING.md#target-release-types)
  - [ ] N/A

- [X] Check this box to confirm: **I have read the [_Submitting Changes_](../blob/main/CONTRIBUTING.md#submitting-changes) section of [`CONTRIBUTING.md`](../blob/main/CONTRIBUTING.md) and reviewed the following information:**

* **SHOULD** update `ChangeLog.md` file(s) appropriately
    * For SDK-based development mode, update `src/{{SERVICE}}/{{SERVICE}}/ChangeLog.md`.
        * A snippet outlining the change(s) made in the PR should be written under the `## Upcoming Release` header in the past tense. 
    * For autorest-based development mode, include the changelog in the PR description.
    * Should **not** change `ChangeLog.md` if no new release is required, such as fixing test case only.
* **SHOULD** regenerate markdown help files if there is cmdlet API change. [Instruction](../blob/main/documentation/development-docs/help-generation.md#updating-all-markdown-files-in-a-module)
* **SHOULD** have proper test coverage for changes in pull request.
* **SHOULD NOT** adjust version of module manually in pull request

### CODE REVIEW

**Review by grizzlytheodore — COMMENTED** [review]
code change look LGTM with assumption that the test runs fine. 
I see build is failing in CI, can that be resolved?

File: src/Compute/Compute.Autorest/test/Invoke-AzSpotPlacementRecommender.Tests.ps1 [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+                desiredSizes = $desiredSizes;
+                desiredCount = 100;
+                desiredLocations = $desiredLocations;
+                availabilityZones = $true
+            }
+            Invoke-AzSpotPlacementRecommender -Location eastus -SpotPlacementRecommenderInput $spotPlacementRecommenderInput -verbose
+        }
+    }
+
+    It 'PostViaIdentityExpanded' -skip {
```
Comment (vidai-msft) [inline src/Compute/Compute.Autorest/test/Invoke-AzSpotPlacementRecommender.Tests.ps1:45]: If you don't plan to write test for this parameter set, please remove the whole section. But we highly recommend testing each parameter set.

File: src/Compute/Compute.Autorest/test/Invoke-AzSpotPlacementRecommender.Tests.ps1 [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+        { throw [System.NotImplementedException] } | Should -Not -Throw
+    }
+
+    It 'PostViaIdentity' -skip {
```
Comment (vidai-msft) [inline src/Compute/Compute.Autorest/test/Invoke-AzSpotPlacementRecommender.Tests.ps1:49]: If you don't plan to write test for this parameter set, please remove the whole section. But we highly recommend testing each parameter set.

**Comment by vidai-msft** [comment]
@Elina-L Please kindly find the following action items.
1. You selected the target release as Public Preview, which means the code will not be merged into main and released in the Az. Instead, we will create a new branch to merge the code and release with a separate package. Please confirm if this is expected.
2. Invoke-AzSpotPlacementRecommender.Recording.json has not been checked in so that the test won't be run in CI.
3. Examples, docs and test files are missing for Get-AzSpotPlacementRecommender.

File: src/Compute/Compute.Autorest/test/Invoke-AzSpotPlacementRecommender.Tests.ps1 (hunk 1 above)
Comment (Elina-L) [reply src/Compute/Compute.Autorest/test/Invoke-AzSpotPlacementRecommender.Tests.ps1]: Removed PostViaIdentityExpanded example

File: src/Compute/Compute.Autorest/test/Invoke-AzSpotPlacementRecommender.Tests.ps1 (hunk 2 above)
Comment (Elina-L) [reply src/Compute/Compute.Autorest/test/Invoke-AzSpotPlacementRecommender.Tests.ps1]: Removed PostViaIdentity test

**Comment by Elina-L** [comment]
> @Elina-L Please kindly find the following action items.
> 
> 1. You selected the target release as Public Preview, which means the code will not be merged into main and released in the Az. Instead, we will create a new branch to merge the code and release with a separate package. Please confirm if this is expected.
> 2. Invoke-AzSpotPlacementRecommender.Recording.json has not been checked in so that the test won't be run in CI.
> 3. Examples, docs and test files are missing for Get-AzSpotPlacementRecommender.

Hi Vincent,

I have a few questions:
1. What is the difference between Az and the separate package? Does that mean users will need to install a separate PS package in order to run the cmdlets? Will the cmdlets of the separate package be automatically be released in the main Az in the future after some time or will there need to be another PR to move it from the separate package to the main package?

2. I am somehow not able to generate the recording files. After running -Record, I see the following:
image
However, no files are generated at the path: 
image
Am I missing some step or are my test files written incorrectly?

3. Removed Get-AzSpotPlacementRecommender references, will not be including that cmdlet in this PR.

**Comment by vidai-msft** [comment]
@Elina-L Public Preview as the name implies, this is a preview package instead of official release. In PSGallery, this package will be suffixed with `-preview` and not included in Az. We generally consider that it contains preview or experimental features that would like our customers to have a try. The package won't be included into Az automatically unless the code is manually ported into main branch. BTW, whether the API version is preview has no relationship with the target release.

**Comment by vidai-msft** [comment]
@Elina-L The autorest generated code is based on the swagger files specified in README.md. If that file is not ready, then codegen pipeline will fail. We can only run that after the swagger is ready.

**Comment by Elina-L** [comment]
> @Elina-L The autorest generated code is based on the swagger files specified in README.md. If that file is not ready, then codegen pipeline will fail. We can only run that after the swagger is ready.

Hi Vincent, the swagger is ready and already merged to main branch: https://github.com/Azure/azure-rest-api-specs/blob/main/specification/compute/resource-manager/Microsoft.Compute/DiagnosticRP/preview/2024-03-01-preview/diagnostic.json

However, the swagger file is pointing to an API version that is currently still being rolled out to Prod ARM manifest in this CRP rollout:  [Compute-CPlat-Core-Manifest-Master - Release-329 - Pipelines (azure.com)](https://dev.azure.com/msazure/One/_releaseProgress?releaseId=12813803&_a=release-pipeline-progress)

Currently, if I run autorest on the remote swagger doc, I can successfully generate the cmdlets and successfully run "build-module.ps1". I can also successfully initiate "run-module.ps1". However, I hit an error when I try to invoke my cmdlet, it fails because the ARM Manifest with version 2024-03-01-preview has not yet been rolled out.

```
$response = Invoke-AzSpotPlacementRecommender -Location eastus -DesiredCount 1 -DesiredLocation $desiredLocations -DesiredSize $desiredSizes
Invoke-AzSpotPlacementRecommender_PostExpanded: No registered resource provider found for location 'eastus' and API version '2024-03-01-preview' for type 'locations/diagnostics'. The supported api-versions are '2021-06-01-preview'. The supported locations are 'westcentralus, southeastasia, eastus2, eastus, southcentralus, westeurope, northeurope, westus2, uksouth, australiaeast, southindia, centralindia, westindia, koreacentral, koreasouth, jioindiawest, japanwest, japaneast, eastasia, canadacentral, canadaeast, southafricanorth, centralus, westus3, westus, ukwest, francecentral, uaenorth, australiacentral, australiasoutheast, switzerlandnorth, germanywestcentral, norwayeast, northcentralus, brazilsouth, swedencentral, eastus2euap, centraluseuap'.
```

Is there any way to hardcode this PR so that the package generated by PS team calls the 2021-06-01-preview API endpoint? There are no errors to generate/build/run the cmdlet, but if it's pointing to the new API which hasn't yet rolled out, some test cases may fail. Would that be an issue to get this PR merged?

**Comment by vidai-msft** [comment]
@Elina-L Code generation is just based on the swagger while the real command execution depends on the ARM manifest as you mentioned.
@grizzlytheodore could you please help check and confirm if public preview is desired as you are more familiar with our process. :-) Thanks!

**Comment by grizzlytheodore** [comment]
@Elina-L i suggest public preview for the Az release. as your REST API calls are also public preview. 
Before you release, you will need to make sure ARM manifest is rollout. We shouldn't distribute cmdlet that does not work yet. 

If you'd like to call older version that works currently, you can update the swagger reference in the readme to call your older version

**Comment by vidai-msft** [comment]
@Elina-L The cut off date for this sprint is your EOD Monday. If you plan to release this feature on Apr 2 with general release, please check in the recording file for test case and leave a comment here as change log message. Thanks!

**Comment by Elina-L** [comment]
> @Elina-L The cut off date for this sprint is your EOD Monday. If you plan to release this feature on Apr 2 with general release, please check in the recording file for test case and leave a comment here as change log message. Thanks!

Hi, I have updated the PR with the recording file for the test case, kindly requesting another PR review.

**Comment by vidai-msft** [comment]
> > @Elina-L The cut off date for this sprint is your EOD Monday. If you plan to release this feature on Apr 2 with general release, please check in the recording file for test case and leave a comment here as change log message. Thanks!
> 
> Hi, I have updated the PR with the recording file for the test case, kindly requesting another PR review.

@Elina-L As per our last communication, the target release should be `General release` instead of `Public preview`. Can you please confirm if this is still the desired option? If so, can you please switch the option in this PR?

**Comment by Elina-L** [comment]
@vidai-msft Thanks, I have updated it to General Release

**Comment by Nickcandy** [comment]
@Elina-L Hi, please leave your changelog here and I will merge it to main branch

**Pull request merged**